### DISCUSSION TITLE
Full page transition issues with AlpineJS + BarbaJS

### DISCUSSION THREAD

**Discussion by 13Groszy** [alpinejs/alpine#4305] [1. Help]
Hi all,

The issue I encounter right now is quite simple, yet I can't find the source of the problem and how to fix it.
I have very basic full page transition applied (covered by barba) with this code:

`barba.init({
    transitions: [{
        name: 'opacity-transition',
        leave(data) {
            return gsap.to(data.current.container, {
                    opacity: 0,
                    duration: 1,
            })
        },
        enter(data) {
            return gsap.from(data.next.container, {
                opacity: 0,
                duration: 1,
            })
        }
    }],
});`

Alpine code:

`document.addEventListener('alpine:init', () => {
    Alpine.data('animatedNumber', () => ({
        startValue: 0,
        endValue: 1,
    }))
});`

Problem: 
When I navigate from url A to url B the transition will be fine, but my alpine code will not be initialised: 
`Alpine Expression Error: animatedNumber is not defined
Expression: "animatedNumber"`

If I use following code in enter(data) before returning animation:
`Alpine.discoverUninitializedComponents((el) => {
                Alpine.initializeComponent(el);
            });`
It seems to initialise components properly, but the animation stop and it looks like first half (leave) take place and second half (enter) is skipped. This literally looks like fading out previous page and refresh on next page.

Does anyone seen this issue before? Can I initialise alpine components without refreshing the page and carry on with animation?
Maybe someone used better library to animate full page transition on websites which use Alpine?

TIA
Matt

**Comment by ekwoka** [comment]
So, the alpine init event won't rerun when you swap the new code in.

Also, can you clarify if you're using Livewire or something?

`discoverUninitializedComponents` and `initializeComponent` are not methods on the Alpine global.

But you just need to evaluate the script tags before changing the DOM. Maybe even just use the actual browser view transitions api.

**Reply by SimoTod** [reply]
They were ages ago (V2)

**Comment by 13Groszy** [comment]
I'm not using Livewire, the stack I use is Orchard Core with Liquid(Shopify) support.

Here's simplified Layout file which might explain a bit more:

`<!DOCTYPE html>`
`<html lang="{{ Culture.Name }}">`
    `<head>`
        `<script defer src="{{ "/dist/scripts.js" | append_version }}"></script>`
        `<script defer src="{{ "/dist/components/swiper.js" | append_version }}"></script>`
    `</head>`
    `<body data-barba="wrapper">`
         `{% render "Areas/Theme/Views/Shared/Header" %}`
        `<main  class="flex flex-col flex-1 is-cover" data-barba="container">`
            `{% render_body %}`
        `</main>`
        `{% resources type: "FootScript" %}`
    `</body>`
`</html>`


So based on that my script's should rerun after page change.

**Reply by SimoTod** [reply]
https://barba.js.org/docs/getstarted/markup/#DOM-structure says Barba only swaps the container (main tag). I assume alpine is not in the main tag and, anyway, alpine"init is only fired once when you load the page the first time while Barba is just swapping part of the page.

You'll need to use the hooks of your tool (https://barba.js.org/docs/advanced/hooks/) to init the new components. Bear in mind that, from what you posted above, it looks like you are using a really old version of Alpine (4y old) so you need to check the old documentation. It's also harder for people to help since Alpine v2 was quite different.

**Reply by 13Groszy** [reply]
So looks like my issue is not with Alpine, but with how it works after barba swap the container. 
Btw. Alpine I use is up to date, only the methods I used here are from Alpine v2. 

Thanks, I'll do more research and if I find solution, I'll pass it here for others.

**Reply by SimoTod** [reply]
: 👍 
Maybe double check the version to make it sure. Those methods don't exist in v3. If they run without errors, you somehow have the old version injected in your page (or perhaps both).

**Reply by ekwoka** [reply]
Yeah, those methods don't exist so you should get `discoverUninitializedComponents is not a function`

**Comment by Britnell** [comment]
Hey there,
im having the same issues with ViewTransition (in browser or astro polyfill)

is there any more follow up info on this? i feel like there are a lot of cases with client-side loading of html with alpine, what is the suggested way to handle this?

i made a simple [stacklbitz demo here](https://stackblitz.com/~/github.com/Britnell/astronomical?initialPath=/alpine/client/a) 

i tried listening for the `alpine:init` as well as `page-transition` events, and this works, but there are still issues with alpine attaching itself to the dom. is there a particular timing or order required here?

```
    document.addEventListener("alpine:init", () => {
      window.setupalpinecomp();
    });
    document.addEventListener("astro:page-load", () => {
      window.setupalpinecomp();
    });
```

i attached the `Alpine.data(...)` functions straight to window, 
else if the page is visited twice we get a `comp already declared error`

you can see by the debug logs i added that the init function runs, but still alpine throws errors that the fields are undefined

**Reply by Britnell** [reply]
about the demo, should load on `/alpine/client/a`
 `/a` has a controlled input, `/b` a counter,
both work on page load, 
but if you navigate over to the other via the link (astro viewtrasitions handles as client side)
then they don't work / load properly.

**Comment by Britnell** [comment]
I just found someone [solve it this way](https://brianschiller.com/blog/2024/05/03/alpinejs-with-astro/)

with a wrapper that checks page status & if the sub components are defined before calling them ... but that seems unnecessarily  complex and verbose, isn't there a prescribed way to init alpine 'components'  (or whatever you call `Alpine.data`) after client side routing ?

**Comment by flimshaw** [comment]
I am also trying to solve this problem. So far, I am getting away with running Alpine.start() in the barba.enter hook. 
```
enter(data) {
    // Initialize Alpine only for the new container's elements
    Alpine.start();
    return fadeAnimation(data.next.container, 'in');
},
```

This seems to work properly, however I am getting a warning from Alpine that .start() should only be run once. Is there a way to initialize new Alpine components in specific container or something more targeted?

**Comment by zz40404991** [comment]
I searched many pages and finally found that the solution at the end of this link is effective. https://github.com/withastro/astro/issues/11133
@13Groszy
